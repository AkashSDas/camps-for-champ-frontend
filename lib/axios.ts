import axios, { AxiosError, AxiosRequestConfig } from "axios";

export var axiosInstance = axios.create({
  baseURL: process.env.NEXT_PUBLIC_BACKEND_URL,
  withCredentials: true,
});

async function fetchFromAPI(URL: string, config: AxiosRequestConfig) {
  try {
    var response = await axiosInstance(URL, config);
  } catch (err) {
    if (err instanceof AxiosError) {
      if (err.response) {
        var error: any = err.response;
      } else if (err.message == "Network Error") {
        var error: any = {
          status: 500,
          data: { message: "Network Error" },
        };
      }
    } else {
      throw new Error("Error occurred while making request to the backend");
    }
  }

  return {
    statusCode: response!?.status ?? error?.status ?? 500,
    data: response!?.data ?? error?.data ?? {},
    error: error?.data,
    success: response!?.status < 300,
  };
}

export default fetchFromAPI;
