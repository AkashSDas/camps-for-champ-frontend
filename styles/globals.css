@import url("https://fonts.googleapis.com/css2?family=Poppins:wght@400;500;600;700;800&display=swap");

.orange-text-gradient {
  background: -webkit-linear-gradient(#df8531, #df6531);
  background-clip: text;
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

.icon > .sidebar-icon {
  fill: #f0f0f0;
}

.icon:hover > .sidebar-icon {
  fill: #df6531;
}

.icon > .sidebar-icon-stroke {
  stroke: #f0f0f0;
}

.icon:hover > .sidebar-icon-stroke {
  stroke: #df6531;
}
