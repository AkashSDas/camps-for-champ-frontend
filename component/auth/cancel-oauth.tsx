import { toast } from "react-hot-toast";
import { useMutation } from "react-query";

import { Text } from "@chakra-ui/react";

import { useUser } from "../../lib/hooks";
import { queryClient } from "../../lib/react-query";
import { cancelOAuth } from "../../services/auth";

export default function CancelOAuthLink() {
  var { accessToken } = useUser();

  var mutation = useMutation({
    mutationFn: () => cancelOAuth(accessToken),
    onSuccess: async () => {
      await queryClient.invalidateQueries(["access-token"]);
      toast.success("Signup cancelled");
    },
    onError: (error: any) => {
      let errorMsg = error?.message;
      if (!errorMsg) toast.error("Something went wrong");
      else {
        if (Array.isArray(errorMsg)) {
          toast.error(error?.message[0] ?? "Something went wrong");
        } else toast.error(error?.message);
      }
    },
  });

  return (
    <Text
      onClick={() => mutation.mutate()}
      cursor="pointer"
      style={{ color: "#1877F2" }}
    >
      Start over
    </Text>
  );
}
