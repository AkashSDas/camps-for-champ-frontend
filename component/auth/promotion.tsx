import { Box, Center, HStack, Image, Text, VStack } from "@chakra-ui/react";

import { theme as customTheme } from "../../lib/chakra";
import { pxToRem } from "../../lib/pxToRem";

export default function Promotion() {
  function Services() {
    return (
      <HStack justifyContent="center" alignItems="center" gap={pxToRem(32)}>
        {["🌍 Solo", "🐶 Pets", "🥮 Food", "🏔️ View"].map((service) => (
          <Center
            key={service}
            fontFamily="heading"
            h={pxToRem(36)}
            px={pxToRem(16)}
            bg="white"
            rounded="full"
          >
            {service}
          </Center>
        ))}
      </HStack>
    );
  }

  return (
    <VStack
      w={pxToRem(660)}
      p={pxToRem(48)}
      rounded="3xl"
      bg={customTheme.colors.brand.lightOrange}
      alignItems="center"
      justifyContent="center"
      gap={pxToRem(32)}
    >
      <Services />

      <Box mixBlendMode="luminosity">
        <Image
          w={pxToRem(548)}
          h={pxToRem(400)}
          src="https://media.giphy.com/media/l0MYC0LajbaPoEADu/giphy.gif"
          alt="Welcome gif"
          objectFit="cover"
          rounded="3xl"
        />
      </Box>

      <Text fontSize="xl" fontWeight="semibold">
        Adventure is allowed
      </Text>
    </VStack>
  );
}
