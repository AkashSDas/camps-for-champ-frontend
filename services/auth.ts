import { AxiosRequestConfig } from "axios";

import { SignupInput } from "../lib/schema";
import fetchFromAPI from "./";

function fetchFromAuth(URL: string, config: AxiosRequestConfig) {
  return fetchFromAPI(`/auth/${URL}`, config);
}

export async function signup(input: SignupInput) {
  var response = await fetchFromAuth("signup", {
    method: "POST",
    data: input,
  });

  if (response.status == 201) {
    return {
      success: true,
      message: response.data.message,
      user: response.data.user,
      accessToken: response.data.accessToken,
    };
  }

  return { success: false, error: response.error };
}

export async function getNewAccessToken() {
  var response = await fetchFromAuth("access-token", { method: "GET" });
  if (response.status == 200) {
    return {
      success: response.status,
      user: response.data.user,
      accessToken: response.data.accessToken,
    };
  }

  return { success: false, error: response.error };
}
