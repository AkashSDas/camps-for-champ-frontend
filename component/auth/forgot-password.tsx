import NextLink from "next/link";
import { useForm } from "react-hook-form";
import toast from "react-hot-toast";

import { Button, FormControl, FormErrorMessage, FormLabel, Heading, Input, Text, VStack } from "@chakra-ui/react";
import { yupResolver } from "@hookform/resolvers/yup";

import { pxToRem } from "../../lib/pxToRem";
import { ForgotPasswordInput, forgotPasswordSchema } from "../../lib/schema";
import { forgotPassword } from "../../services/auth";

export default function ForgotPasswordSection() {
  var { reset, register, handleSubmit, formState } =
    useForm<ForgotPasswordInput>({
      defaultValues: { email: "" },
      resolver: yupResolver(forgotPasswordSchema),
    });

  async function onSubmit(data: ForgotPasswordInput) {
    try {
      var response = await forgotPassword(data);

      if (response.success) {
        toast.success(response.message);
        reset();
      } else {
        let error = response.error;
        let errorMsg = error?.message;
        if (!errorMsg) toast.error("Something went wrong");
        else {
          if (Array.isArray(errorMsg)) {
            toast.error(error?.message[0] ?? "Something went wrong");
          } else toast.error(error?.message);
        }
      }
    } catch (error) {
      toast.error("Something went wrong");
    }
  }

  return (
    <VStack justifyContent="center" gap={pxToRem(32)} w={pxToRem(400)}>
      <Heading fontSize={pxToRem(40)}>Forgot Password</Heading>

      <Text textAlign="center" fontWeight="500">
        Password reset link will be sent to your registered email address
      </Text>

      {/* Forgot password form */}
      <VStack
        w="full"
        as="form"
        justifyContent="center"
        gap={pxToRem(24)}
        onSubmit={handleSubmit((data) => onSubmit(data))}
      >
        <FormControl isInvalid={formState.errors.email ? true : false}>
          <FormLabel>Email</FormLabel>
          <Input
            type="email"
            variant="base"
            {...register("email")}
            borderColor={
              formState.errors.email
                ? "red.500"
                : formState.touchedFields.email
                ? "green.500"
                : "#CDCDCD"
            }
          />

          <FormErrorMessage>{formState.errors.email?.message}</FormErrorMessage>
        </FormControl>

        <Button
          variant="regularSolid"
          type="submit"
          disabled={formState.isSubmitting}
        >
          {formState.isSubmitting ? "Loading..." : "Send instructions"}
        </Button>
      </VStack>

      <Text fontWeight="medium">
        {"Don't"} have an account?{" "}
        <NextLink href="/auth/signup" style={{ color: "#1877F2" }}>
          Signup
        </NextLink>
      </Text>
    </VStack>
  );
}
