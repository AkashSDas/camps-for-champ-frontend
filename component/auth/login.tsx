import NextLink from "next/link";
import { useRouter } from "next/router";
import { useForm } from "react-hook-form";
import toast from "react-hot-toast";

import { Button, FormControl, FormErrorMessage, FormLabel, Heading, HStack, Input, Text, VStack } from "@chakra-ui/react";
import { yupResolver } from "@hookform/resolvers/yup";

import { pxToRem } from "../../lib/pxToRem";
import { LoginInput, loginSchema } from "../../lib/schema";
import { login } from "../../services/auth";
import { FacebookIcon, GoogleIcon, TwitterIcon } from "../icons/social";

export default function LoginSection() {
  var router = useRouter();
  var { reset, register, handleSubmit, formState } = useForm<LoginInput>({
    defaultValues: { email: "", password: "" },
    resolver: yupResolver(loginSchema),
  });

  async function onSubmit(data: LoginInput) {
    var response = await login(data);
    if (!response.success) {
      let errorMsg = response.error?.message;
      if (!errorMsg) toast.error("Something went wrong");
      else {
        if (Array.isArray(errorMsg)) {
          toast.error(response.error?.message[0] ?? "Something went wrong");
        } else toast.error(response.error?.message);
      }
    } else {
      toast.success("Login successful");
      reset();
      router.push("/");
    }
  }

  return (
    <VStack justifyContent="center" gap={pxToRem(32)} w={pxToRem(400)}>
      <Heading fontSize={pxToRem(40)}>Login</Heading>

      <VStack justifyContent="center" gap={pxToRem(24)} width="full">
        {/* OAuth buttons */}
        <HStack gap={pxToRem(32)}>
          <Button variant="outline">
            <GoogleIcon />
          </Button>

          <Button variant="outline">
            <TwitterIcon />
          </Button>

          <Button variant="outline">
            <FacebookIcon />
          </Button>
        </HStack>

        <Heading size="md">OR</Heading>

        {/* Signup form */}
        <VStack
          w="full"
          as="form"
          justifyContent="center"
          gap={pxToRem(24)}
          onSubmit={handleSubmit(onSubmit)}
        >
          <FormControl isInvalid={formState.errors.email ? true : false}>
            <FormLabel>Email</FormLabel>
            <Input
              type="email"
              variant="base"
              {...register("email")}
              borderColor={
                formState.errors.email
                  ? "red.500"
                  : formState.touchedFields.email
                  ? "green.500"
                  : "#CDCDCD"
              }
            />

            <FormErrorMessage>
              {formState.errors.email?.message}
            </FormErrorMessage>
          </FormControl>

          <FormControl isInvalid={formState.errors.password ? true : false}>
            <FormLabel>Password</FormLabel>
            <Input
              type="password"
              variant="base"
              {...register("password")}
              borderColor={
                formState.errors.password
                  ? "red.500"
                  : formState.touchedFields.password
                  ? "green.500"
                  : "#CDCDCD"
              }
            />

            <FormErrorMessage>
              {formState.errors.password?.message}
            </FormErrorMessage>
          </FormControl>

          <Button
            variant="regularSolid"
            type="submit"
            disabled={formState.isSubmitting}
          >
            {formState.isSubmitting ? "Loading..." : "Login with email"}
          </Button>
        </VStack>

        <Text fontWeight="medium">
          {"Don't"} have an account?{" "}
          <NextLink href="/auth/signup" style={{ color: "#1877F2" }}>
            Signup
          </NextLink>
        </Text>
      </VStack>
    </VStack>
  );
}
