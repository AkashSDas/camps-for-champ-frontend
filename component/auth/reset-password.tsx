import NextLink from "next/link";
import { useRouter } from "next/router";
import { useForm } from "react-hook-form";
import toast from "react-hot-toast";
import { useMutation } from "react-query";

import { Button, FormControl, FormErrorMessage, FormLabel, Heading, HStack, Input, Text, VStack } from "@chakra-ui/react";
import { yupResolver } from "@hookform/resolvers/yup";

import { useUser } from "../../lib/hooks";
import { pxToRem } from "../../lib/pxToRem";
import { queryClient } from "../../lib/react-query";
import { LoginInput, loginSchema, ResetPasswordInput, resetPasswordSchema } from "../../lib/schema";
import { login, resetPassword } from "../../services/auth";
import { FacebookIcon, GoogleIcon, TwitterIcon } from "../icons/social";

export default function ResetPasswordSection() {
  var router = useRouter();
  var { reset, register, handleSubmit, formState } =
    useForm<ResetPasswordInput>({
      defaultValues: { password: "", confirmPassword: "" },
      resolver: yupResolver(resetPasswordSchema),
    });

  async function onSubmit(data: ResetPasswordInput) {
    try {
      var response = await resetPassword(data, router.query.token as string);

      if (response.success) {
        toast.success(response.message);
        reset();
      } else {
        let error = response.error;
        let errorMsg = error?.message;
        if (!errorMsg) toast.error("Something went wrong");
        else {
          if (Array.isArray(errorMsg)) {
            toast.error(error?.message[0] ?? "Something went wrong");
          } else toast.error(error?.message);
        }
      }
    } catch (error) {
      toast.error("Something went wrong");
    }
  }

  return (
    <VStack justifyContent="center" gap={pxToRem(32)} w={pxToRem(400)}>
      <Heading fontSize={pxToRem(40)}>Reset Password</Heading>

      {/* Reset password form */}
      <VStack
        w="full"
        as="form"
        justifyContent="center"
        gap={pxToRem(24)}
        onSubmit={handleSubmit((data) => onSubmit(data))}
      >
        <FormControl isInvalid={formState.errors.password ? true : false}>
          <FormLabel>Password</FormLabel>
          <Input
            type="password"
            variant="base"
            {...register("password")}
            borderColor={
              formState.errors.password
                ? "red.500"
                : formState.touchedFields.password
                ? "green.500"
                : "#CDCDCD"
            }
          />

          <FormErrorMessage>
            {formState.errors.password?.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl
          isInvalid={formState.errors.confirmPassword ? true : false}
        >
          <FormLabel>Confirm password</FormLabel>
          <Input
            type="password"
            variant="base"
            {...register("confirmPassword")}
            borderColor={
              formState.errors.confirmPassword
                ? "red.500"
                : formState.touchedFields.confirmPassword
                ? "green.500"
                : "#CDCDCD"
            }
          />

          <FormErrorMessage>
            {formState.errors.confirmPassword?.message}
          </FormErrorMessage>
        </FormControl>

        <Button
          variant="regularSolid"
          type="submit"
          disabled={formState.isSubmitting}
        >
          {formState.isSubmitting ? "Loading..." : "Update password"}
        </Button>
      </VStack>

      <Text fontWeight="medium">
        Already have an account?{" "}
        <NextLink href="/auth/login" style={{ color: "#1877F2" }}>
          Login
        </NextLink>
      </Text>
    </VStack>
  );
}
