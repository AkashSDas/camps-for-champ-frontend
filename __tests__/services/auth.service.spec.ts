import axios from "axios";
import MockAdapter from "axios-mock-adapter";

import { describe, expect, it } from "@jest/globals";

import { axiosInstance } from "../../lib/axios";
import { signup } from "../../services/auth.service";

// This sets the mock adapter on the default instance
var mock = new MockAdapter(axiosInstance, { onNoMatch: "throwException" });

describe("Auth services", () => {
  describe("signup", () => {
    describe("when the user is created successfully", () => {
      it("the return the user and access token", async () => {
        mock.onPost("/auth/email-signup").reply(201, {
          user: userStub(),
          accessToken: accessTokenStub(),
        });

        const result = await signup({
          email: userStub().email,
          password: "testingTEST123@",
        });

        expect(result).toMatchSnapshot();
        expect(result.accessToken).toBeDefined();
        expect(result.user).toBeDefined();
      });
    });

    describe("when the user already exist", () => {
      it("then get already exists error", async () => {
        mock.onPost("/auth/email-signup").reply(400, {
          statusCode: 400,
          message: "User already exists",
          error: "Bad Request",
        });

        const result = await signup({
          email: userStub().email,
          password: "testingTEST123@",
        });

        expect(result).toMatchSnapshot();
        expect(result.message).toEqual("User already exists");
      });
    });

    describe("network error", () => {
      it("then get already exists error", async () => {
        mock.onPost("/auth/email-signup").networkError();

        const result = await signup({
          email: userStub().email,
          password: "testingTEST123@",
        });

        expect(result).toMatchSnapshot();
        expect(result.message).toEqual("Network Error");
      });
    });
  });
});

function userStub() {
  return {
    email: "james@gmail.com",
    active: false,
    banned: false,
    verified: false,
    roles: ["base"],
    _id: "63bbc5597413cc38dacdaf59",
    oauthProviders: [],
    userId: "acc_XLL2Wp8C0k0eZWpY",
    createdAt: "2023-01-09T07:42:17.784Z",
    updatedAt: "2023-01-09T07:42:17.784Z",
    __v: 0,
  };
}

function accessTokenStub() {
  return "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImphbWVzQGdtYWlsLmNvbSIsImlhdCI6MTY3MzI1MDEzNywiZXhwIjoxNjczMjUwNDM3fQ.u4ulNIxOfYKqjYFWUQzXoqWb0YCkKxRVxfPeHvDOI6k";
}
