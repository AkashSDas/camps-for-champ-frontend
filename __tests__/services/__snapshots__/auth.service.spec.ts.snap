// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth services signup network error then get already exists error 1`] = `
{
  "message": "Network Error",
  "success": false,
}
`;

exports[`Auth services signup when the user already exist then get already exists error 1`] = `
{
  "message": "User already exists",
  "success": false,
}
`;

exports[`Auth services signup when the user is created successfully the return the user and access token 1`] = `
{
  "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImphbWVzQGdtYWlsLmNvbSIsImlhdCI6MTY3MzI1MDEzNywiZXhwIjoxNjczMjUwNDM3fQ.u4ulNIxOfYKqjYFWUQzXoqWb0YCkKxRVxfPeHvDOI6k",
  "message": "Account created successfully",
  "success": true,
  "user": {
    "__v": 0,
    "_id": "63bbc5597413cc38dacdaf59",
    "active": false,
    "banned": false,
    "createdAt": "2023-01-09T07:42:17.784Z",
    "email": "james@gmail.com",
    "oauthProviders": [],
    "roles": [
      "base",
    ],
    "updatedAt": "2023-01-09T07:42:17.784Z",
    "userId": "acc_XLL2Wp8C0k0eZWpY",
    "verified": false,
  },
}
`;
